### PR TITLE
Extension exposing and cleanup

### PR DESCRIPTION
**Pull request by LeoDog896**
In order for other plugins to manage extensions better, they need to know more about them, and this exposes the internal DiscoveredExtension object instead of the old ExtensionDescription.

### CODE REVIEW

**Review by jglrxavpok — APPROVED** [review]
Looks good to me

File: src/main/java/net/minestom/server/extensions/DiscoveredExtension.java [hunk 1]
```diff
@@ -81,7 +120,27 @@ public File getOriginalJar() {
...
-    static void verifyIntegrity(@NotNull DiscoveredExtension extension) {
+    public MinestomExtensionClassLoader removeMinestomExtensionClassLoader() {
+        MinestomExtensionClassLoader oldClassLoader = getMinestomExtensionClassLoader();
+        setMinestomExtensionClassLoader(null);
+        return oldClassLoader;
+    }
+
+    public void setMinestomExtensionClassLoader(@Nullable MinestomExtensionClassLoader minestomExtensionClassLoader) {
+        this.minestomExtensionClassLoader = minestomExtensionClassLoader;
+    }
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/extensions/DiscoveredExtension.java:131]: Why make these functions public? That's opening a way for users to break how extension classloaders work, and someone will complain some day.

File: src/main/java/net/minestom/server/extras/selfmodification/MinestomExtensionClassLoader.java [hunk 2]
```diff
@@ -105,8 +105,14 @@ protected void finalize() throws Throwable {
...
+
+        if (mainClassName.equals(name))
             return true;
-        return children.stream().anyMatch(c -> c.isMainExtensionClass(name));
+
+        for (MinestomExtensionClassLoader child : children) {
+            if (child.isMainExtensionClass(name)) return true;
+        }
+
+        return false;
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/extras/selfmodification/MinestomExtensionClassLoader.java:116]: While I don't mind the change, I'm curious as to why you changed this to something that has the same semantic?

File: src/main/java/net/minestom/server/extras/selfmodification/MinestomRootClassLoader.java [hunk 3]
```diff
@@ -298,11 +296,7 @@ public boolean loadModifier(URL[] urls, String codeModifierClass) {
             }
             return true;
         } catch (ClassNotFoundException | InvocationTargetException | InstantiationException | IllegalAccessException | NoSuchMethodException e) {
-            if(MinecraftServer.getExceptionManager() != null) {
-                MinecraftServer.getExceptionManager().handleException(e);
-            } else {
-                e.printStackTrace();
-            }
+            MinecraftServer.getExceptionManager().handleException(e);
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/extras/selfmodification/MinestomRootClassLoader.java:299]: I might have missed the change previously, but the exception manager is not guaranteed to be present at this stage. Especially if you are loading a class before `MinecraftServer#init` is called.

**Review by jglrxavpok — CHANGES_REQUESTED** [review]

File: src/main/java/net/minestom/server/extras/selfmodification/MinestomExtensionClassLoader.java (hunk 2 above)
Comment (LeoDog896) [reply src/main/java/net/minestom/server/extras/selfmodification/MinestomExtensionClassLoader.java:116]: I'm personally not a big fan of streams, though I'm not sure where streams stand in Minestom -- like we mainly use for loops but theres streams (mainly for testing / debugging) so I wanted to follow that with this.

File: src/main/java/net/minestom/server/extras/selfmodification/MinestomRootClassLoader.java (hunk 3 above)
Comment (LeoDog896) [reply src/main/java/net/minestom/server/extras/selfmodification/MinestomRootClassLoader.java:299]: _Did not notice that_ Ill patch that really quickly then.

**Review by jglrxavpok — APPROVED** [review]

File: src/main/java/net/minestom/server/extras/selfmodification/MinestomExtensionClassLoader.java (hunk 2 above)
Comment (TheMode) [reply src/main/java/net/minestom/server/extras/selfmodification/MinestomExtensionClassLoader.java:116]: I believe that the stream in this case is cleaner

**Pull request merged**